File: simulation.py
#By Dameesh Kumar
import random
import math
import numpy as np

# Test data - a 14-day forecast
forecast = [5, 3, 7, 2, 6, 4, 8, 1, 9, 3, 5, 2, 7, 4]

print("Original forecast:", forecast)
print("Number of days:", len(forecast))

# Poisson simulation with NumPy Library
def simulate_poisson_demand(forecast):
    return np.random.poisson(forecast).tolist()


poisson_result = simulate_poisson_demand(forecast)
print("\nPoisson simulation results:")
print("Forecast: ", forecast)
print("Poisson:  ", poisson_result)

# Binomial simulation with NumPy Library
def simulate_binomial_demand(forecast, trials=10):
    probabilities = np.minimum(np.array(forecast) / trials, 1.0)
    return np.random.binomial(trials, probabilities).tolist()


binomial_result = simulate_binomial_demand(forecast, trials=10)
print("\nBinomial simulation results:")
print("Forecast: ", forecast)
print("Binomial: ", binomial_result)

def compare_simulations(forecast, trials=10, num_runs=5):
    print(f"\nComparing Poisson vs Binomial simulations ({num_runs} runs):")
    print("Forecast:", forecast)
    print("-" * 80)
    
    for run in range(num_runs):
        poisson_result = simulate_poisson_demand(forecast)
        binomial_result = simulate_binomial_demand(forecast, trials)
        
        print(f"Run {run + 1}:")
        print(f"  Poisson:  {poisson_result}")
        print(f"  Binomial: {binomial_result}")
        print()


compare_simulations(forecast, trials=10, num_runs=5)

def analyze_simulations(forecast, trials=10, num_runs=1000):
    """
    Perform statistical analysis comparing Poisson vs Binomial simulations
    """
    print(f"\nStatistical Analysis ({num_runs} runs):")
    print("=" * 60)
    
    # Store all results for analysis
    poisson_errors = []
    binomial_errors = []
    poisson_totals = []
    binomial_totals = []
    
    # Run many simulations
    for _ in range(num_runs):
        poisson_result = simulate_poisson_demand(forecast)
        binomial_result = simulate_binomial_demand(forecast, trials)
        
        # Calculate errors (difference from forecast)
        poisson_error = sum(abs(p - f) for p, f in zip(poisson_result, forecast))
        binomial_error = sum(abs(b - f) for b, f in zip(binomial_result, forecast))
        
        poisson_errors.append(poisson_error)
        binomial_errors.append(binomial_error)
        poisson_totals.append(sum(poisson_result))
        binomial_totals.append(sum(binomial_result))
    
    # Calculate statistics
    forecast_total = sum(forecast)
    
    print(f"Original forecast total: {forecast_total}")
    print()
    
    # Poisson Statistics
    print("POISSON RESULTS:")
    print(f"  Average total demand: {np.mean(poisson_totals):.2f}")
    print(f"  Standard deviation: {np.std(poisson_totals):.2f}")
    print(f"  Average absolute error: {np.mean(poisson_errors):.2f}")
    print(f"  Error standard deviation: {np.std(poisson_errors):.2f}")
    
    print()
    
    # Binomial Statistics  
    print("BINOMIAL RESULTS:")
    print(f"  Average total demand: {np.mean(binomial_totals):.2f}")
    print(f"  Standard deviation: {np.std(binomial_totals):.2f}")
    print(f"  Average absolute error: {np.mean(binomial_errors):.2f}")
    print(f"  Error standard deviation: {np.std(binomial_errors):.2f}")
    
    print()
    
    # Comparison
    print("COMPARISON:")
    poisson_better = sum(1 for p, b in zip(poisson_errors, binomial_errors) if p < b)
    binomial_better = sum(1 for p, b in zip(poisson_errors, binomial_errors) if b < p)
    
    print(f"  Poisson more accurate: {poisson_better}/{num_runs} times ({poisson_better/num_runs*100:.1f}%)")
    print(f"  Binomial more accurate: {binomial_better}/{num_runs} times ({binomial_better/num_runs*100:.1f}%)")
    
    if np.mean(poisson_errors) < np.mean(binomial_errors):
        print(f"  Winner: Poisson (lower average error)")
    else:
        print(f"  Winner: Binomial (lower average error)")


analyze_simulations(forecast, trials=10, num_runs=1000)

File: simulation.kt
// By Dameesh Kumar
import kotlin.math.*
import kotlin.random.Random

fun main() {
    // Test data - a 14-day forecast
    val forecast = listOf(5, 3, 7, 2, 6, 4, 8, 1, 9, 3, 5, 2, 7, 4)
    
    println("Original forecast: $forecast")
    println("Number of days: ${forecast.size}")
    
    // Poisson simulation
    val poissonResult = simulatePoissonDemand(forecast)
    println("\nPoisson simulation results:")
    println("Forecast: $forecast")
    println("Poisson:  $poissonResult")
    
    // Binomial simulation
    val binomialResult = simulateBinomialDemand(forecast, trials = 10)
    println("\nBinomial simulation results:")
    println("Forecast: $forecast")
    println("Binomial: $binomialResult")
    
    // Compare simulations
    compareSimulations(forecast, trials = 10, numRuns = 5)
    
    // Analyze simulations
    analyzeSimulations(forecast, trials = 10, numRuns = 1000)
}

// Poisson simulation function
fun simulatePoissonDemand(forecast: List<Int>): List<Int> {
    return forecast.map { lambda ->
        generatePoisson(lambda.toDouble())
    }
}

// Binomial simulation function
fun simulateBinomialDemand(forecast: List<Int>, trials: Int = 10): List<Int> {
    return forecast.map { f ->
        val probability = minOf(f.toDouble() / trials, 1.0)
        generateBinomial(trials, probability)
    }
}

// Compare simulations function
fun compareSimulations(forecast: List<Int>, trials: Int = 10, numRuns: Int = 5) {
    println("\nComparing Poisson vs Binomial simulations ($numRuns runs):")
    println("Forecast: $forecast")
    println("-".repeat(80))
    
    repeat(numRuns) { run ->
        val poissonResult = simulatePoissonDemand(forecast)
        val binomialResult = simulateBinomialDemand(forecast, trials)
        
        println("Run ${run + 1}:")
        println("  Poisson:  $poissonResult")
        println("  Binomial: $binomialResult")
        println()
    }
}

// Statistical analysis function
fun analyzeSimulations(forecast: List<Int>, trials: Int = 10, numRuns: Int = 1000) {
    println("\nStatistical Analysis ($numRuns runs):")
    println("=".repeat(60))
    
    // Store all results for analysis
    val poissonErrors = mutableListOf<Int>()
    val binomialErrors = mutableListOf<Int>()
    val poissonTotals = mutableListOf<Int>()
    val binomialTotals = mutableListOf<Int>()
    
    // Run many simulations
    repeat(numRuns) {
        val poissonResult = simulatePoissonDemand(forecast)
        val binomialResult = simulateBinomialDemand(forecast, trials)
        
        // Calculate errors (difference from forecast)
        val poissonError = poissonResult.zip(forecast) { p, f -> abs(p - f) }.sum()
        val binomialError = binomialResult.zip(forecast) { b, f -> abs(b - f) }.sum()
        
        poissonErrors.add(poissonError)
        binomialErrors.add(binomialError)
        poissonTotals.add(poissonResult.sum())
        binomialTotals.add(binomialResult.sum())
    }
    
    // Calculate statistics
    val forecastTotal = forecast.sum()
    
    println("Original forecast total: $forecastTotal")
    println()
    
    // Poisson Statistics
    println("POISSON RESULTS:")
    println("  Average total demand: ${"%.2f".format(poissonTotals.average())}")
    println("  Standard deviation: ${"%.2f".format(poissonTotals.standardDeviation())}")
    println("  Average absolute error: ${"%.2f".format(poissonErrors.average())}")
    println("  Error standard deviation: ${"%.2f".format(poissonErrors.standardDeviation())}")
    
    println()
    
    // Binomial Statistics
    println("BINOMIAL RESULTS:")
    println("  Average total demand: ${"%.2f".format(binomialTotals.average())}")
    println("  Standard deviation: ${"%.2f".format(binomialTotals.standardDeviation())}")
    println("  Average absolute error: ${"%.2f".format(binomialErrors.average())}")
    println("  Error standard deviation: ${"%.2f".format(binomialErrors.standardDeviation())}")
    
    println()
    
    // Comparison
    println("COMPARISON:")
    val poissonBetter = poissonErrors.zip(binomialErrors) { p, b -> p < b }.count { it }
    val binomialBetter = poissonErrors.zip(binomialErrors) { p, b -> b < p }.count { it }
    
    println("  Poisson more accurate: $poissonBetter/$numRuns times (${"%.1f".format(poissonBetter.toDouble() / numRuns * 100)}%)")
    println("  Binomial more accurate: $binomialBetter/$numRuns times (${"%.1f".format(binomialBetter.toDouble() / numRuns * 100)}%)")
    
    if (poissonErrors.average() < binomialErrors.average()) {
        println("  Winner: Poisson (lower average error)")
    } else {
        println("  Winner: Binomial (lower average error)")
    }
}

// Utility function to generate Poisson random numbers
fun generatePoisson(lambda: Double): Int {
    val L = exp(-lambda)
    var k = 0
    var p = 1.0
    
    do {
        k++
        p *= Random.nextDouble()
    } while (p > L)
    
    return k - 1
}

// Utility function to generate Binomial random numbers
fun generateBinomial(n: Int, p: Double): Int {
    var count = 0
    repeat(n) {
        if (Random.nextDouble() < p) {
            count++
        }
    }
    return count
}

// Extension function to calculate standard deviation
fun List<Int>.standardDeviation(): Double {
    val mean = this.average()
    val variance = this.map { (it - mean).pow(2) }.average()
    return sqrt(variance)
}